lib/minitest/assertions.rb

```diff
@@ -464,25 +464,28 @@ def capture_io
     def capture_subprocess_io
       _synchronize do
         begin
-          require "tempfile"
-
-          captured_stdout, captured_stderr = Tempfile.new("out"), Tempfile.new("err")
+          stdout_reader, stdout_writer = IO.pipe
+          stderr_reader, stderr_writer = IO.pipe
 
           orig_stdout, orig_stderr = $stdout.dup, $stderr.dup
-          $stdout.reopen captured_stdout
-          $stderr.reopen captured_stderr
 
-          yield
+          begin
+            $stdout.reopen stdout_writer
+            $stderr.reopen stderr_writer
+
+            yield
+          ensure
+            $stdout.reopen orig_stdout
+            $stderr.reopen orig_stderr
+          end
 
-          $stdout.rewind
-          $stderr.rewind
+          stdout_writer.close
+          stderr_writer.close
 
-          return captured_stdout.read, captured_stderr.read
+          return stdout_reader.read, stderr_reader.read
```

Comment:
There is a bug: a pipe only has a limited capacity, if the output to stdout or stderr exceeds it, this will block indefinitely because we only start reading after *all* the writing is done.

The solution is to start reading in a thread.

Reply:
Maybe specify a length to read and do it in a loop until empty? There's probably 100 ways to do this and most are bad. :P

Reply:
This is the fix I used in another project.

``` ruby
            stdout_reader_thread = Thread.new { stdout_reader.read }
            stderr_reader_thread = Thread.new { stderr_reader.read }

            yield
          ensure
            $stdout.reopen orig_stdout
            $stderr.reopen orig_stderr
          end

          stdout_writer.close
          stderr_writer.close

          return stdout_reader_thread.join.value, stderr_reader_thread.join.value
```

I pushed a commit to address this issue.